### DISCUSSION TITLE
Nested Secrets Dir Support?

### DISCUSSION THREAD

**Discussion by 0xdade** [pydantic/pydantic#2928] [Question]
Hey all,

I'm trying to build a parent Settings object, which has many settings underneath it that are namespaced into feature sets, basically. So `Settings()` actually consists almost entirely of other settings objects. Some of these sub-settings require loading secrets, from `secrets_dir`, but the naming of the variable in the settings doesn't match up with the secrets dir.

Example:

```python
class S3Settings(BaseSettings):
    """
    S3Settings represents the collection of settings necessary for configuring our connection to an S3 compatible object store.

    This maps to supported fields in the toml [s3] section.
    """
    endpoint: str
    bucket: str
    access_key: Secret
    secret_key: Secret
    use_tls: bool

    class Config:
        env_prefix = 's3_'
        secrets_dir = '/secrets/s3'

class Settings(BaseSettings):
    """All application settings belong here."""

    secret_key: Secret

    redis_dsn: RedisDsn = "redis://user:pass@localhost:6379/1"
    pg_dsn: PostgresDsn = "postgres://user:pass@localhost:5432/foobar"

    s3: S3Settings = S3Settings()

    auth: AuthSettings = AuthSettings()
    brand: BrandSettings = BrandSettings()

      class Config:
          """Tells pydantic where and how to load settings."""

            env_file_encoding = "utf-8"
            env_file = ".env"
            secrets_dir = "/secrets"

            @classmethod
            def customise_sources(
                cls,
                init_settings: SettingsSourceCallable,
                env_settings: SettingsSourceCallable,
                file_secret_settings: SettingsSourceCallable,
            ):
                """Injects `config.toml` into the Settings sources."""
                return (
                    init_settings,
                    toml_config_settings_source,
                    env_settings,
                    file_secret_settings,
                )

```

I've been searching and searching and am struggling to find a solution, and haven't been able to make sense of pydantic's `env_settings.py` file well enough to understand what I could patch myself to make this work.

I found [this PR](https://github.com/samuelcolvin/pydantic/issues/1727), which seems to want to do a similar thing but with env prefixes. Unfortunately it seems that the env_prefix doesn't apply to the file names in the secrets_dir, and I can't seem to correctly scope secrets_dir for S3Settings above to be `/secrets/s3`. So `/secrets/s3_access_key` won't work with the env_prefix, and `/secrets/s3/access_key` won't work with custom secrets_dir being defined in the Config of S3Settings. I've tried many variations of this and haven't had success.

I know I could just extract the s3 secret_key and s3 access_key into the top level Settings, and name them s3_secret_key and s3_access_key, but I'd like to be able to access all relevant S3 settings via `Settings().s3.access_key` and `Settings().s3.secret_key`

Is there a best practice I could use for this? Maybe what I want is just bad, and there is a more standardized best practice I could adopt instead. Otherwise, is there a supported way of doing this now? Or is there an unsupported but patchable way of doing it now?

Thanks!

**Comment by gawsoftpl** [comment]
#### If you add env_prefix to Config the Secret file name should have this prefix.

remove env_prefix from S3 Config.

```sh
# tree /secrets

/secrets
└── s3
    └── access_key
    └── secret_key
```

should be:
```python
from pydantic import BaseSettings, SecretStr

class S3Settings(BaseSettings):
    endpoint: str
    bucket: str
    access_key: SecretStr
    secret_key: SecretStr

    class Config:
        secrets_dir = '/secrets/s3'

class Settings(BaseSettings):
    class Config:
        secrets_dir = '/secrets/'

    s3: S3Settings = S3Settings()
```
Now in client
```python
settings = Settings()
settings.s3.secret_key.get_secret_value()
```

**Comment by 0xdade** [comment]
Ah the problem is that I want the env prefixes to also be scoped. An env var of just "SECRET_KEY" isn't very useful. I want the env vars to be prefixed with a unique prefix but don't want to have to name those settings with that prefix.
